components/idenpub/idenpub.go

```diff
@@ -0,0 +1,184 @@
+package idenpub
+
+import (
+	"bytes"
+	"math"
+	"sync"
+
+	"github.com/iden3/go-iden3-core/db"
+	"github.com/iden3/go-iden3-core/merkletree"
+)
+
+var (
+	cacheIKey  = []byte("cacheI")
+	cltRootKey = []byte("cltRoot")
+	rotRootKey = []byte("rotRoot")
+	retRootKey = []byte("retRoot")
+	rotKey     = []byte("rot")
+	retKey     = []byte("ret")
+)
+
+// IdenPub is a interface, that for the moment will be satisfied at least by IdenPubHTTP & IdenPubIPFS.
+type IdenPub interface {
+	Publish()
+}
+
+// IdenPubHTTP satisfies the IdenPub interface, and stores in a leveldb the published RoT & ReT to be returned when requested.
+type IdenPubHTTP struct {
+	sync.RWMutex
+	db  db.Storage
+	rot *merkletree.MerkleTree
+	ret *merkletree.MerkleTree
+}
+
+// AddLeafRoT adds a new leaf to the given MerkleTree, which contains the Root
+func AddLeafRoT(mt *merkletree.MerkleTree, root merkletree.Hash) error {
+	l := NewLeafRoT(root)
+	return mt.AddEntry(l.Entry())
+}
+
+// AddLeafReT adds a new leaf to the given MerkleTree, which contains the Nonce & Version
+func AddLeafReT(mt *merkletree.MerkleTree, nonce, version uint32) error {
+	l := NewLeafReT(nonce, version)
+	return mt.AddEntry(l.Entry())
+}
+
+// NewIdenPubHTTP returns a new IdenPubHTTP
+func NewIdenPubHTTP(db db.Storage, rot *merkletree.MerkleTree, ret *merkletree.MerkleTree) *IdenPubHTTP {
+	return &IdenPubHTTP{
+		db:  db,
+		rot: rot,
+		ret: ret,
+	}
+}
+
+// Publish publishes the RoT and ReT to the configured way of publishing
+func (i *IdenPubHTTP) Publish(claimsRoot, rootsRoot, revocationsRoot *merkletree.Hash) error {
+	// RoT
+	w := bytes.NewBufferString("")
+	err := i.rot.DumpTreeIoWriter(w, rootsRoot)
```

Comment:
Notice that the leaves of the RootsRoots are 32 bytes while the laves of the DumpTreeIoWriter are 32 * 8 bytes in hex (*2)!

Reply:
At this moment has been putted in bytes to string without hex, a further task is to implement a way to crop the Roots from inside the Entries without affecting the rest of the nodes that are stored (detecting by the first byte type)

Reply:
#288